stream-rpc/src/connection.rs

```diff
@@ -0,0 +1,389 @@
+        error: Option<JsonRpcError>,
+    ) -> anyhow::Result<()> {
+        counters::INVALID_REQUESTS
+            .with_label_values(&[
+                self.connection_context.transport.as_str(),
+                method.map_or("unknown", |m| m.as_str()),
+                error.as_ref().map_or("unknown", |e| e.code_as_str()),
+                self.connection_context.sdk_info.language.as_str(),
+                &self.connection_context.sdk_info.version.to_string(),
+            ])
+            .inc();
+        self.send_error(id, error).await
+    }
+
+    pub fn unsubscribe(&self, id: &RequestIdentifier) -> Option<()> {
+        self.running_tasks.stop(id)
+    }
+
+    pub fn is_closed(&self) -> bool {
+        return self.sender.is_closed();
+    }
+
+    pub async fn received_message(&self, db: Arc<dyn DbReader>, message: String) {
+        match JsonRpcRequest::from_str(&message) {
+            Ok(request) => {
+                debug!(
+                    logging::SubscriptionRequestLog {
+                        transport: self.connection_context.transport.as_str(),
+                        remote_addr: self.connection_context.remote_addr.clone(),
+                        client_id: self.id,
+                        rpc_method: Some(request.method_name()),
+                    },
+                    "subscription request"
+                );
+                if let Err(err) = self.handle_rpc_request(db, &request) {
+                    self.send_error_with_metrics(
+                        Some(request.method_request.method()),
+                        Some(request.id),
+                        Some(err),
+                    )
+                        .await
+                        .ok();
+                }
+            }
+            Err((err, method, id)) => {
+                // We couldn't parse the request- it's not valid json or an unknown structure
+                debug!(
+                    logging::SubscriptionRequestLog {
+                        transport: self.connection_context.transport.as_str(),
+                        remote_addr: self.connection_context.remote_addr.clone(),
+                        client_id: self.id,
+                        rpc_method: method.map_or(None, |v| Some(v.as_str())),
+                    },
+                    "failed to parse subscription request ({})", &err
+                );
+                metric_subscription_rpc_received(
+                    &self,
+                    method.map_or("unknown", |m| m.as_str()),
+                    counters::RpcResult::Error,
+                );
+                self.send_error_with_metrics(method, id, Some(err))
+                    .await
+                    .ok();
+            }
+        }
+    }
+
+    fn handle_rpc_request(
+        &self,
+        db: Arc<dyn DbReader>,
+        request: &JsonRpcRequest,
+    ) -> Result<(), JsonRpcError> {
+        let mut tasks = self.running_tasks.tasks.lock().unwrap();
```

Comment:
lock? will it cause sequence call to handle_rpc_request?

Reply:
what do you mean by `sequence call`?

I need to add documentation on this, but: 
- The lock is within the scope of one client (subscribing, unsubscribing, disconnecting, or some combination therein)
- The `call_method` is responsible for doing validation on the parameters, and returning a `Option<JoinHandle>` (tokio task) if they launched a subscription
- The lock must be held until we can determine whether or not we have a subscription because otherwise there is a race condition if a user submits the same rpc id (`RequestIdentifier`) again after we’ve verified it’s not used, but before we insert it, which would result in losing track of that subscription task (and leaking green threads)

Does that help answer the question?